src/retry.ts

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { errorMessage } from './errors';
+import { sleep } from './time';
+
+/**
+ * RetryOptions are the config options for the withRetry function
+ */
+export interface RetryOptions {
+  attempts?: number;
+  backoff?: number;
+}
+
+/**
+ * withRetry a simple implementation of a retry mechanism with fibonacci backoff
+ *
+ * @param fn. A function to retry on failure
+ * @param opts. The retry options
+ * @throws {Error}
+ */
+export async function withRetry<T>(fn: () => Promise<T>, opts?: RetryOptions): Promise<T> {
+  const DEFAULT_ATTEMPTS = 3;
+  const DEFAULT_BACKOFF_MILLISECONDS = 30 * 1000; // 30 seconds
+
+  let attemptLimit = opts?.attempts || DEFAULT_ATTEMPTS;
+  let currentBackoff = opts?.backoff || DEFAULT_BACKOFF_MILLISECONDS;
+  let prevBackoff = 0;
+
+  // recursive execution function
+  const execRetry = async (): Promise<T> => {
+    try {
+      return await fn.call(null);
```

Comment:
Hmm - this could be unexpected because `this.foo` inside the given function wouldn't resolve anymore.